rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, reverting a sway program only produces an opaque error code.
+Sometimes this code is unusual enough that with the source code of the contract
+it is possible to figure out what is going on, but a lot of the time it's a
+meaningless number that only means that some failure happened somewhere.
+
+It's not reasonable to expect people to be able to debug sway programs in these
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings minimally burden the final bytecode size.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Error types
+
+Error types are special enums that allow developers to signal that each variant
+is associated with an error message.
+
+An error type implements the `AbiError` trait, but it is not recomended that you
+do so manually.
+
+Instead, to create a new error type, use the following syntax and annotations on
+an enum:
+
+```sway
+#[abi_error]
+enum MyError {
+  #[error("error A")]
+  A: (),
+  #[error("error B")]
+  B: (u64, u8),
+}
+```
+
+This will automatically implement the `AbiError` marker trait for this enum
+and let the compiler know what error messages correspond to what variants of
+the enum.
+
+It will also populate the ABI specification file with error messages in the
+metadata of the enum type. Allowing users to check the meaning of an error
+without the need to store and decode error messages on chain.
+
+For instance, the following contract method may return a `MyError` variant which
+will be compact but can still be associated with an error message:
+
+```sway
+fn do_something(self) -> Result<(), MyError> {
+  // ...
+  return Err(MyError::A); // user will know that this means "error A"
+  // ...
+}
+```
+
+## `panic`
+
+When encountering an irrecoverable error in a sway program, it is customary to
+revert and produce an informative error code.
+
+The recommended way of doing this is to use the the `panic` intrinsic.
+
+`panic` is a compiler intrinsic that will
+produce a unique revert code for each of its invocations and populate the ABI
+specification file with a list of such codes and corresponding information about
+where it was invoked and with what kind of arguments.
+
+If all you want is an error message, you can panic with a string literal:
+
+```sway
```

Comment:
From a consumer perspective if we have to accommodate string literal panics as well as error type panics one could for see inconsistent error handling patterns emerging. My preference would solely Error types.

Reply:
Agree with this; at the end it may come down to what is most optimal (gas-wise); for example on ethereum explicit errors are generally more gas-efficient than reverts

Reply:
On one side I agree and like the idea of a uniform approach. On the other side, insisting only on `Error` would create a terrible boilerplate for cases (often cases), when a single error description is just enough.

Instead of just having `panic("This is my issue.");` devs would need to do all the ceremony:
```
#[abi_error]
enum DummyError {
   #[error("This is my issue.")]
   Dummy: (),
}

panic(DummyError::Dummy);
```

Honestly, I think we will soon end up with much more inconsistencies around the choice of `Dummy` word :smile: 

That's why I think supporting `str` is a good tradeof.

Reply:
Thinking through again, the argument of simply panic'ing out is also just as valid; especially for quick iteration

Reply:
Thanks for the clarification @IGI-111 and @ironcev , agree on both points so resolving this.